storage/replica_command.go

```diff
@@ -250,6 +250,10 @@ func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args
 	if args.Txn == nil {
 		return reply, nil, util.Errorf("no transaction specified to EndTransaction")
 	}
+	if !bytes.Equal(args.Key, args.Txn.Key) {
```

Comment:
Is it possible for us to check this immediately on receiving the request. That way you need to check it only once.